### PR TITLE
Implement 3-Beam Solution for ADCP

### PR DESCRIPTION
**Pull request by cwingard**
In response to Redmine ticket 14302, this update to the adcp_beam2ins function in the adcp_functions.py module will support the calculation of velocity data products in RCA ADCPs when one of the 4 beams has failed or is compromised.

Will require an update to Preload to add additional inputs to the function, but does not change the outputs.

### CODE REVIEW

File: ion_functions/data/perf/test_adcp_performance.py [hunk 1]
```diff
@@ -77,7 +86,7 @@ def test_adcp_beam_eastward(self):
         z = np.repeat(self.depth, 10000)
         dt = np.repeat(self.ntp, 10000)
 
-        self.profile(stats, af.adcp_beam_eastward, b1, b2, b3, b4, h, p, r, vf, lat, lon, z, dt)
+        self.profile(stats, af.adcp_beam_eastward, b1, b2, b3, b4, pg1, pg2, pg3, pg4, h, p, r, vf, lat, lon, dt)
```
Comment (krosburg) [inline ion_functions/data/perf/test_adcp_performance.py:88]: is z no longer needed here?

File: ion_functions/data/adcp_functions.py [hunk 2]
```diff
@@ -717,47 +746,54 @@ def vadcp_beam_vertical_true(b1, b2, b3, b4, b5, h, p, r, vf):
...
+        b1 = beam 1 velocity profiles in beam coordinates (VELTURB-B1_L0) [mm s-1]
+        b2 = beam 2 velocity profiles in beam coordinates (VELTURB-B2_L0) [mm s-1]
+        b3 = beam 3 velocity profiles in beam coordinates (VELTURB-B3_L0) [mm s-1]
+        b4 = beam 4 velocity profiles in beam coordinates (VELTURB-B4_L0) [mm s-1]
+        b5 = beam 5 velocity profiles in beam coordinates (VELTURB-B5_L0) [mm s-1]
+        pg1 = percent good estimate for beam 1 [percent]
+        pg2 = percent good estimate for beam 2 [percent]
+        pg3 = percent good estimate for beam 3 [percent]
+        pg4 = percent good estimate for beam 4 [percent]
+        pg5 = percent good estimate for beam 4 [percent]
```
Comment (sgfoote) [inline ion_functions/data/adcp_functions.py]: Make pg5 be the percent good estimate for beam 5, not 4 (i.e. copy paste error duplicated the pg4 description).

File: ion_functions/data/adcp_functions.py [hunk 3]
```diff
@@ -859,50 +901,90 @@ def adcp_beam2ins(b1, b2, b3, b4):
...
-        u, v, w, e = adcp_beam2ins(b1, b2, b3, b4)
+        x, y, z, e = adcp_beam2ins(b1, b2, b3, b4, pg1, pg2, pg3, pg4)
 
             where
 
-        u = "east" velocity profiles in instrument coordinates [mm s-1]
-        v = "north" velocity profiles in instrument coordinates [mm s-1]
-        w = "vertical" velocity profiles in instrument coordinates [mm s-1]
-        e = "error" velocity profiles [mm s-1]
+        x = x axis velocity profiles in instrument coordinates [mm s-1]
```
Comment (sgfoote) [inline ion_functions/data/adcp_functions.py:913]: Could we retain some indication that our x axis here is "east-west" and our y axis "north-south" and the z axis "vertical"? While that's the intuitive interpretation, it seems better to explicitly identify what the coordinates mean.

File: ion_functions/data/perf/test_adcp_performance.py [hunk 4]
```diff
@@ -25,19 +25,23 @@ class TestADCPPerformance(PerformanceTestCase):
...
-        self.b4 = np.array([-0.2160, -0.6050, -0.0920, -0.0580,  0.4840,
-                            -0.0050,  0.3380,  0.1750, -0.0800, -0.5490]) * 1000
-        self.echo = np.array([0, 25, 50, 75, 100, 125, 150, 175, 200, 225, 250])
+        # set test inputs -- values from DPS (reset to integers, matches format from instrument)
+        self.b1 = np.array([[-30, -295, -514, -234, -188, 203, -325, 305, -204, -294]])
+        self.b2 = np.array([[180, -132, 213, 309, 291, 49, 188, 373, 2, 172]])
+        self.b3 = np.array([[-398, -436, -131, -473, -443, 188, -168, 291, -179, 8]])
+        self.b4 = np.array([[-216, -605, -92, -58, 484, -5, 338, 175, -80, -549]])
+
+        self.pg1 = np.array([[100, 100, 100,  24, 100, 100,  24, 100, 100, 100]])
```
Comment (sgfoote) [inline ion_functions/data/perf/test_adcp_performance.py:34]: Should we include some values of 25 or 26 in these percent good arrays just to exercise the other side of the percent good floor (with something closer to the threshold than 100)?

**Review by sgfoote — APPROVED** [review]

**Review by krosburg — APPROVED** [review]

File: ion_functions/data/adcp_functions.py (hunk 2 above)
Comment (cwingard) [reply ion_functions/data/adcp_functions.py]: Done

File: ion_functions/data/adcp_functions.py (hunk 3 above)
Comment (cwingard) [reply ion_functions/data/adcp_functions.py:913]: No, that is incorrect. At this stage, the x, y, z axis are relative to the instrument and require a further transformation (utilizing the compass, tilt/roll sensors, and the instrument orientation) to align them with the east-west/north-south axis. This function is specifically to transform the data from the individual beam coordinates to instrument coordinates.

File: ion_functions/data/perf/test_adcp_performance.py (hunk 4 above)
Comment (cwingard) [reply ion_functions/data/perf/test_adcp_performance.py:34]: Mixed up the percent good array to include a bunch of other values. Hopefully that will work.

File: ion_functions/data/perf/test_adcp_performance.py (hunk 1 above)
Comment (cwingard) [reply ion_functions/data/perf/test_adcp_performance.py:88]: Deleted z from the function.

**Review by steinermg — APPROVED** [review]

**Pull request merged**